gpcontrib/gp_orca/gporca.c

```diff
@@ -19,10 +91,34 @@ _PG_init(void)
 
 	if (!IS_QUERY_DISPATCHER())
 		return;
+
+	if (Gp_role == GP_ROLE_DISPATCH)
+	{
+		/* Initialize GPOPT */
+		OptimizerMemoryContext = AllocSetContextCreate(TopMemoryContext,
+													"GPORCA Top-level Memory Context",
+													ALLOCSET_DEFAULT_MINSIZE,
+													ALLOCSET_DEFAULT_INITSIZE,
+													ALLOCSET_DEFAULT_MAXSIZE);
+
+		InitGPOPT();
+
+		prev_planner = planner_hook;
+		planner_hook = gp_orca_planner;
+	}
 }
 
 void
 _PG_fini(void)
 {
+	if (Gp_role == GP_ROLE_DISPATCH)
+	{
+		planner_hook = prev_planner;
+
+		TerminateGPOPT();
+
+		if (OptimizerMemoryContext != NULL)
+			MemoryContextDelete(OptimizerMemoryContext);
+	}
```

Comment:
The `ShutdownPostgres` function (which previously called the `TerminateGPOPT` function) is called before the process terminates (`before_shmem_exit`), while the `_PG_fini` function is never called at all and has been removed from recent versions of `Postgres`.